src/librustc_driver/target_features.rs

```diff
@@ -0,0 +1,85 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use syntax::{ast, attr};
+use rustc::session::Session;
+use syntax::parse::token::InternedString;
+use syntax::parse::token::intern_and_get_ident as intern;
+
+pub fn add_configuration(cfg: &mut ast::CrateConfig, sess: &Session) {
+    let tf = InternedString::new("target_feature");
+    macro_rules! fillout {
+        ($($func: ident, $name: expr;)*) => {{
+            $(if $func(sess) {
+                cfg.push(attr::mk_name_value_item_str(tf.clone(), intern($name)))
+            })*
+        }}
+    }
+    fillout! {
+        has_sse, "sse";
+        has_sse2, "sse2";
+        has_sse3, "sse3";
+        has_ssse3, "ssse3";
+        has_sse41, "sse4.1";
+        has_sse42, "sse4.2";
+        has_avx, "avx";
+        has_avx2, "avx2";
+        has_neon, "neon";
+        has_vfp, "vfp";
+    }
+}
+
+
+fn features_contain(sess: &Session, s: &str) -> bool {
+    sess.target.target.options.features.contains(s) ||
+        sess.opts.cg.target_feature.contains(s)
```

Comment:
I believe you can [get ahold of a `TargetMachineRef`](https://github.com/rust-lang/rust/blob/master/src/librustc_trans/back/write.rs#L166), get ahold of the [`MCSubtargetInfo`](https://github.com/llvm-mirror/llvm/blob/8ae0800ae15d7016819cb1d96f6540c3a63a28c9/include/llvm/Target/TargetMachine.h#L160), get the [list of features](https://github.com/llvm-mirror/llvm/blob/8ae0800ae15d7016819cb1d96f6540c3a63a28c9/include/llvm/MC/MCSubtargetInfo.h#L69-L73), and then test for [various features](https://github.com/llvm-mirror/llvm/blob/8ae0800ae15d7016819cb1d96f6540c3a63a28c9/lib/Target/X86/X86.td#L30-L200) being available.

It's probably not super critical that this happens immediately though.